File: codes/lua/011.operator_spread.lua
JSON = (loadfile "utils/JSON.lua")() -- Thanks to Jeffrey Friedl's awesome work, checkout his awesome personal blog at http://regex.info/blog/lua/json

function s_print(...)
    local parameters = {...}
    local result = ""
    for _, parameter in ipairs(parameters) do
        result = result .. tostring(parameter)
    end
    print(result)
end

-- There's no JavaScript-like Spread Syntax (...) in Lua.
-- But, we can create our own function to mimic it in Lua.

function pretty_json_stringify(anything) return JSON:encode_pretty(anything, 'etc', { pretty=true, indent="    ", array_newline=true }) end

function pretty_array_of_primitives(an_array_of_primitives)
    local result = "["
    for array_item_index, array_item in ipairs(an_array_of_primitives) do
        if ((type(array_item) ~= "string") and (type(array_item) ~= "number") and (type(array_item) ~= "boolean") and (array_item ~= "nil")) then
            goto next
        end
        if (array_item == "nil") then
            result = result .. "null"
        end
        if ((type(array_item) == "string") and (array_item ~= "nil")) then
            result = result .. "\"" .. array_item .. "\""
        end
        if (type(array_item) == "number") then
            result = result .. array_item
        end
        if (type(array_item) == "boolean") then
            result = result .. tostring(array_item)
        end
        if (array_item_index ~= #an_array_of_primitives) then
            result = result .. ", "
        end
        ::next::
    end
    result = result .. "]"
    return result
end

function get_type(anything)
    if (type(anything) ~= "table") then
       return type(anything) 
    end
    if (next(anything) == nil) then
        return "empty_table"
    end
    for k, v in pairs(anything) do
        if ((type(k) == "number") and ((k >= 1) and (k <= #anything))) then
            return "array"
        end
    end
    return "object"
end

function get_object_length(an_object)
    local object_length = 0
    for object_key, object_value in pairs(an_object) do
        object_length = object_length + 1
    end
    return object_length
end

spread_syntax_object = function(...)
    local parameters = {...}
    local new_object = {}
    for parameter_index, parameter in ipairs(parameters) do
        local parameter_type = get_type(parameter)
        if (parameter_type == "object") then
            for object_key, object_value in pairs(parameter) do
                new_object[object_key] = object_value
            end
            goto next
        end
        if (parameter_type == "array") then
            for array_item_index, array_item in ipairs(parameter) do
                new_object[tostring(array_item_index)] = array_item
            end
            goto next
        end
        ::next::
    end
    return new_object
end

spread_syntax_array = function(...)
    local parameters = {...}
    local new_array = {}
    for parameter_index, parameter in ipairs(parameters) do
        local parameter_type = get_type(parameter)
        if (parameter_type == "object") then
            local object_length = get_object_length(parameter)
            if (object_length == 1) then
                for object_key, object_value in pairs(parameter) do
                    table.insert(new_array, object_value)
                end
                goto next
            end
            table.insert(new_array, parameter)
            goto next
        end
        if (parameter_type == "array") then
            for array_item_index, array_item in ipairs(parameter) do
                table.insert(new_array, array_item)
            end
            goto next
        end
        ::next::
    end
    return new_array
end

print('\n-- JavaScript-like Spread Syntax (...) in Lua')

fruits = {"Mango", "Melon", "Banana"}
s_print("fruits: ", pretty_array_of_primitives(fruits))

vegetables = {"Carrot", "Tomato"}
s_print("vegetables: ", pretty_array_of_primitives(vegetables))

country_capitals_in_asia = {
    Thailand = "Bangkok",
    China = "Beijing",
    Japan = "Tokyo"
}
s_print("country_capitals_in_asia: ", pretty_json_stringify(country_capitals_in_asia))

country_capitals_in_europe = {
    France = "Paris",
    England = "London"
}
s_print("country_capitals_in_europe: ", pretty_json_stringify(country_capitals_in_europe))

print("\n-- [...array1, ...array2]:\n")

combination1 = spread_syntax_array(fruits, vegetables)
s_print("combination1: ", pretty_json_stringify(combination1))
-- combination1: [
--     "Mango",
--     "Melon",
--     "Banana",
--     "Carrot",
--     "Tomato"
-- ]

combination2 = spread_syntax_array(fruits, {"Cucumber", "Onions"})
s_print("combination2: ", pretty_json_stringify(combination2))
-- combination2: [
--     "Mango",
--     "Melon",
--     "Banana",
--     "Cucumber",
--     "Onions"
-- ]

print("\n-- { ...object1, ...object2 }:\n")

combination3 = spread_syntax_object(country_capitals_in_asia, country_capitals_in_europe)
s_print("combination3: ", pretty_json_stringify(combination3))
-- combination3: {
--     "Thailand": "Bangkok",
--     "China": "Beijing",
--     "Japan": "Tokyo",
--     "France": "Paris",
--     "England": "London"
-- }

combination4 = spread_syntax_object(country_capitals_in_asia, { Germany = "Berlin", Italy = "Rome" })
s_print("combination4: ", pretty_json_stringify(combination4))
-- combination4: {
--     "Thailand": "Bangkok",
--     "China": "Beijing",
--     "Japan": "Tokyo",
--     "Germany": "Berlin",
--     "Italy": "Rome"
-- }

print("\n-- [...array1, array2]:\n")

combination5 = spread_syntax_array(fruits, { vegetables = vegetables })
s_print("combination5: ", pretty_json_stringify(combination5))
-- combination5: [
--     "Mango",
--     "Melon",
--     "Banana",
--     [
--         "Carrot",
--         "Tomato"
--     ]
-- ]

combination6 = spread_syntax_array(fruits, { vegetables = {"Cucumber", "Onions"} })
s_print("combination6: ", pretty_json_stringify(combination6))
-- combination6: [
--     "Mango",
--     "Melon",
--     "Banana",
--     [
--         "Cucumber",
--         "Onions"
--     ]
-- ]

print("\n-- [...array1, object1]:\n")

combination7 = spread_syntax_array(fruits, { country_capitals_in_asia = country_capitals_in_asia })
s_print("combination7: ", pretty_json_stringify(combination7))
-- combination7: [
--     "Mango",
--     "Melon",
--     "Banana",
--     {
--         "Thailand": "Bangkok",
--         "China": "Beijing",
--         "Japan": "Tokyo"
--     }
-- ]

combination8 = spread_syntax_array(fruits, { country_capitals_in_europe = { Germany = "Berlin", Italy = "Rome" } })
s_print("combination8: ", pretty_json_stringify(combination8))
-- combination8: [
--     "Mango",
--     "Melon",
--     "Banana",
--     {
--         "Germany": "Berlin",
--         "Italy": "Rome"
--     }
-- ]

print("\n-- { ...object1, object2 }:\n")

combination9 = spread_syntax_object(country_capitals_in_asia, { country_capitals_in_europe = country_capitals_in_europe })
s_print("combination9: ", pretty_json_stringify(combination9))
-- combination9: {
--     "Thailand" : "Bangkok",
--     "China" : "Beijing",
--     "Japan" : "Tokyo",
--     "country_capitals_in_europe" : {
--        "France" : "Paris",
--        "England" : "London"
--     }
--  }

combination10 = spread_syntax_object(country_capitals_in_asia, { country_capitals_in_europe = { Germany = "Berlin", Italy = "Rome" } })
s_print("combination10: ", pretty_json_stringify(combination10))
-- combination10: {
--     "Thailand": "Bangkok",
--     "China": "Beijing",
--     "Japan": "Tokyo",
--     "country_capitals_in_europe": {
--         "Germany": "Berlin",
--         "Italy": "Rome"
--     }
-- }

print("\n-- { ...object1, array2 }:\n")

combination11 = spread_syntax_object(country_capitals_in_asia, { vegetables = vegetables })
s_print("combination11: ", pretty_json_stringify(combination11))
-- combination11: {
--     "Thailand": "Bangkok",
--     "China": "Beijing",
--     "Japan": "Tokyo",
--     "vegetables": [
--         "Carrot",
--         "Tomato"
--     ]
-- }

combination12 = spread_syntax_object(country_capitals_in_asia, { vegetables = {"Cucumber", "Onions"} })
s_print("combination12: ", pretty_json_stringify(combination12))
-- combination12: {
--     "Thailand": "Bangkok",
--     "China": "Beijing",
--     "Japan": "Tokyo",
--     "vegetables": [
--         "Cucumber",
--         "Onions"
--     ]
-- }

print("\n-- { ...object1, ...array2 }:\n")

combination13 = spread_syntax_object(country_capitals_in_asia, vegetables)
s_print("combination13: ", pretty_json_stringify(combination13))
-- combination13: {
--     "Thailand" : "Bangkok",
--     "China" : "Beijing",
--     "Japan" : "Tokyo",
--     "1" : "Carrot",
--     "2" : "Tomato"
--  }

combination14 = spread_syntax_object(country_capitals_in_asia, {"Cucumber", "Onions"})
s_print("combination14: ", pretty_json_stringify(combination14))
-- combination14: {
--     "Thailand" : "Bangkok",
--     "China" : "Beijing",
--     "Japan" : "Tokyo",
--     "1" : "Cucumber",
--     "2" : "Onions"
--  }

-- print("\n-- [...array1, ...object1]: -- this combination throw an error in JavaScript\n")

-- this combination throw an error in JavaScript
-- combination_error_in_javascript1 = spread_syntax_array(fruits, country_capitals_in_asia)
-- s_print("combination_error_in_javascript1: ", pretty_json_stringify(combination_error_in_javascript1))

-- this combination throw an error in JavaScript
-- combination_error_in_javascript2 = spread_syntax_array(fruits, { Germany = "Berlin", Italy = "Rome" })
-- s_print("combination_error_in_javascript2: ", pretty_json_stringify(combination_error_in_javascript2))

File: codes/perl/011_operator_spread.pl
use strict;
# use warnings;

my $js_like_type_ref = {"Null" => "Null", "Boolean" => "Boolean", "String" => "String", "Numeric" => "Numeric", "Object" => "Object", "Array" => "Array", "Function" => "Function"};

sub boolean {
    my ($anything_ref) = @_;
    return ($anything_ref ? (1 == 1) : (1 == 0));
}

sub string {
    my ($anything_ref) = @_;
    return ($anything_ref . "");
}

sub object {
    my ($any_object_ref) = @_;
    return \%{$any_object_ref};
}

sub is_like_js_null {
    my ($anything_ref) = @_;
    return ((defined $anything_ref) ? (1 == 0) : (1 == 1));
}

sub is_like_js_string {
    my ($anything_ref) = @_;
    return (1 == 0) if is_like_js_null($anything_ref);
    return ($anything_ref ne (string($anything_ref) + 0));
}

sub is_triple_equal {
    my ($left_ref, $right_ref) = @_;
    return ($left_ref eq $right_ref) if (is_like_js_string($left_ref) && is_like_js_string($right_ref));
    return ($left_ref == $right_ref);
}

sub is_like_js_numeric {
    my ($anything_ref) = @_;
    return (1 == 0) if is_like_js_null($anything_ref);
    return is_triple_equal(($anything_ref + 0), $anything_ref);
}

sub is_like_js_object {
    my ($anything_ref) = @_;
    return (1 == 0) if is_like_js_null($anything_ref);
    return is_triple_equal((ref $anything_ref), "HASH");
}

sub is_like_js_array {
    my ($anything_ref) = @_;
    return (1 == 0) if is_like_js_null($anything_ref);
    return is_triple_equal((ref $anything_ref), "ARRAY");
}

sub is_like_js_function {
    my ($anything_ref) = @_;
    return (1 == 0) if is_like_js_null($anything_ref);
    return is_triple_equal((ref $anything_ref), "CODE");
}

sub throw_error_if_null {
    my ($anything_ref) = @_;
    die "object key not found in the object" if is_like_js_null($anything_ref);
    return $anything_ref;
}

sub get_type {
    my ($anything_ref) = @_;
    return throw_error_if_null($js_like_type_ref->{"Null"}) if is_like_js_null($anything_ref);
    return throw_error_if_null($js_like_type_ref->{"Function"}) if is_like_js_function($anything_ref);
    return throw_error_if_null($js_like_type_ref->{"Object"}) if is_like_js_object($anything_ref);
    return throw_error_if_null($js_like_type_ref->{"Array"}) if is_like_js_array($anything_ref);
    return throw_error_if_null($js_like_type_ref->{"String"}) if is_like_js_string($anything_ref);
    return throw_error_if_null($js_like_type_ref->{"Numeric"}) if is_like_js_numeric($anything_ref);
    my $anything_ref_unknown_type = (ref $anything_ref);
    return (is_triple_equal($anything_ref_unknown_type, "") ? '"UNKNOWN"' : $anything_ref_unknown_type);
}

sub json_stringify {
    my ($anything_ref, %optional_keyword_argument_object) = @_;
    my $optional_keyword_argument_object_ref = {%optional_keyword_argument_object};
    my $pretty = boolean($optional_keyword_argument_object_ref->{"pretty"});
    my $indent_default = (" " x 4);
    my $indent_level = 0;
    my $json_stringify_inner_ref;
    $json_stringify_inner_ref = sub {
        my ($anything_inner_ref) = @_;
        my $anything_inner_ref_type = get_type($anything_inner_ref);
        return "null" if (is_triple_equal($anything_inner_ref_type, throw_error_if_null($js_like_type_ref->{"Null"})));
        return "[object Function]" if (is_triple_equal($anything_inner_ref_type, throw_error_if_null($js_like_type_ref->{"Function"})));
        if (is_triple_equal($anything_inner_ref_type, throw_error_if_null($js_like_type_ref->{"Object"}))) {
            my $object_keys_ref = [keys(%{$anything_inner_ref})];
            return "{}" if (scalar(@{$object_keys_ref}) == 0);
            $indent_level += 1;
            my $result = ($pretty ? ("{\n" . ($indent_default x $indent_level)) : "{ ");
            my $object_entry_index = 0;
            while (my ($object_key, $object_value) = each(%{$anything_inner_ref})) {
                $result .= '"' . $object_key . '": ' . $json_stringify_inner_ref->($object_value);
                $result .= ($pretty ? (",\n" . ($indent_default x $indent_level)) : ", ") if (($object_entry_index + 1) != scalar(@{$object_keys_ref}));
                $object_entry_index += 1;
            }
            $indent_level -= 1;
            $result .= ($pretty ? ("\n" . ($indent_default x $indent_level) . "}") : " }");
            return $result;
        }
        if (is_triple_equal($anything_inner_ref_type, throw_error_if_null($js_like_type_ref->{"Array"}))) {
            my $array_length = scalar(@{$anything_inner_ref});
            return "[]" if ($array_length == 0);
            $indent_level += 1;
            my $result = ($pretty ? ("[\n" . ($indent_default x $indent_level)) : "[");
            for (my $array_item_index = 0; ($array_item_index < scalar(@{$anything_inner_ref})); $array_item_index += 1) {
                my $array_item = $anything_inner_ref->[$array_item_index];
                $result .= $json_stringify_inner_ref->($array_item);
                $result .= ($pretty ? (",\n" . ($indent_default x $indent_level)) : ", ") if (($array_item_index + 1) != $array_length);
            }
            $indent_level -= 1;
            $result .= ($pretty ? ("\n" . ($indent_default x $indent_level) . "]") : "]");
            return $result;
        }
        return '"' . $anything_inner_ref . '"' if (is_triple_equal($anything_inner_ref_type, throw_error_if_null($js_like_type_ref->{"String"})));
        return string($anything_inner_ref) if (is_triple_equal($anything_inner_ref_type, throw_error_if_null($js_like_type_ref->{"Numeric"})));
        return $anything_inner_ref_type;
    };
    return $json_stringify_inner_ref->($anything_ref);
}

sub array_reduce {
    # JavaScript-like Array.reduce() function
    my ($callback_function_ref, $any_array_ref, $initial_value_ref) = @_;
    my $result_ref = $initial_value_ref;
    for (my $array_item_index = 0; ($array_item_index < scalar(@{$any_array_ref})); $array_item_index += 1) {
        my $array_item_ref = $any_array_ref->[$array_item_index];
        $result_ref = $callback_function_ref->($result_ref, $array_item_ref, $array_item_index, $any_array_ref);
    }
    return $result_ref;
}

sub string_interpolation {
    return (array_reduce((sub {
        my ($current_result_ref, $current_argument_ref) = @_;
        my $current_argument_type = get_type($current_argument_ref);
        return ($current_result_ref . $current_argument_ref) if (is_triple_equal($current_argument_type, throw_error_if_null($js_like_type_ref->{"String"})));
        return ($current_result_ref . (json_stringify($current_argument_ref->[0]))) if (is_triple_equal($current_argument_type, throw_error_if_null($js_like_type_ref->{"Array"})) && (scalar(@{$current_argument_ref}) == 1));
        return ($current_result_ref . (json_stringify($current_argument_ref)));
    }), \@_, ""));
}

sub console_log {
    print(string_interpolation(@_) . "\n");
}

sub optional_chaining {
    my ($anything_ref, @array_index_or_object_key_or_function_argument_array) = @_;
    my $anything_ref_type = get_type($anything_ref);
    return $anything_ref->(@array_index_or_object_key_or_function_argument_array) if (is_triple_equal($anything_ref_type, throw_error_if_null($js_like_type_ref->{"Function"})));
    return $anything_ref if ((!is_triple_equal($anything_ref_type, throw_error_if_null($js_like_type_ref->{"Object"})) && !is_triple_equal($anything_ref_type, throw_error_if_null($js_like_type_ref->{"Array"}))) || (scalar(@array_index_or_object_key_or_function_argument_array) == 0));
    return array_reduce((sub {
        my ($current_result_ref, $current_item_ref) = @_;
        my $current_result_ref_type = get_type($current_result_ref);
        my $current_item_ref_type = get_type($current_item_ref);
        return $anything_ref->{(string($current_item_ref))} if (is_triple_equal($current_result_ref_type, throw_error_if_null($js_like_type_ref->{"Null"})) && is_triple_equal($anything_ref_type, throw_error_if_null($js_like_type_ref->{"Object"})) && is_triple_equal($current_item_ref_type, throw_error_if_null($js_like_type_ref->{"String"})));
        return $anything_ref->[(int($current_item_ref))] if (is_triple_equal($current_result_ref_type, throw_error_if_null($js_like_type_ref->{"Null"})) && is_triple_equal($anything_ref_type, throw_error_if_null($js_like_type_ref->{"Array"})) && is_triple_equal($current_item_ref_type, throw_error_if_null($js_like_type_ref->{"Numeric"})) && (((int($current_item_ref)) >= 0) || ((int($current_item_ref)) == -1)) && (scalar(@{$anything_ref}) > (int($current_item_ref))));
        return $current_result_ref->{(string($current_item_ref))} if (is_triple_equal($current_result_ref_type, throw_error_if_null($js_like_type_ref->{"Object"})) && is_triple_equal($current_item_ref_type, throw_error_if_null($js_like_type_ref->{"String"})));
        return $current_result_ref->[(int($current_item_ref))] if (is_triple_equal($current_result_ref_type, throw_error_if_null($js_like_type_ref->{"Array"})) && is_triple_equal($current_item_ref_type, throw_error_if_null($js_like_type_ref->{"Numeric"})) && (((int($current_item_ref)) >= 0) || ((int($current_item_ref)) == -1)) && (scalar(@{$current_result_ref}) > (int($current_item_ref))));
        return undef;
    }), \@array_index_or_object_key_or_function_argument_array, undef);
}

sub object_from_entries {
    # JavaScript-like Object.fromEntries() function
    my ($any_object_entries_ref) = @_;
    my $new_object_ref = {};
    while (my ($object_entry_index, $object_entry_ref) = each(@{$any_object_entries_ref})) {
        my $object_key = $object_entry_ref->[0];
        my $object_value = $object_entry_ref->[1];
        $new_object_ref->{string($object_key)} = $object_value;
    }
    return $new_object_ref;
}

sub array_entries {
    # JavaScript-like Array.entries() function
    my ($any_array_ref) = @_;
    my $new_object_entries_ref = [];
    for (my $array_item_index = 0; ($array_item_index < scalar(@{$any_array_ref})); $array_item_index += 1) {
        my $array_item = $any_array_ref->[$array_item_index];
        push(@{$new_object_entries_ref}, [$array_item_index, $array_item]);
    }
    return $new_object_entries_ref;
}

sub array_every {
    # JavaScript-like Array.every() function array_every_v4
    my ($callback_function_ref, $any_array_ref) = @_;
    for (my $array_item_index = 0; ($array_item_index < scalar(@{$any_array_ref})); $array_item_index += 1) {
        my $array_item = $any_array_ref->[$array_item_index];
        return (1 == 0) if (!boolean($callback_function_ref->($array_item, $array_item_index, $any_array_ref)));
    }
    return (1 == 1);
}

sub array_filter {
    # JavaScript-like Array.filter() function array_filter_v3
    my ($callback_function_ref, $any_array_ref) = @_;
    my $new_array_ref = [];
    for (my $array_item_index = 0; ($array_item_index < scalar(@{$any_array_ref})); $array_item_index += 1) {
        my $array_item = $any_array_ref->[$array_item_index];
        push(@{$new_array_ref}, $array_item) if (boolean($callback_function_ref->($array_item, $array_item_index, $any_array_ref)));
    }
    return $new_array_ref;
}

sub array_find_index {
    # JavaScript-like Array.findIndex() function array_find_index_v4
    my ($callback_function_ref, $any_array_ref) = @_;
    for (my $array_item_index = 0; ($array_item_index < scalar(@{$any_array_ref})); $array_item_index += 1) {
        my $array_item = $any_array_ref->[$array_item_index];
        return $array_item_index if (boolean($callback_function_ref->($array_item, $array_item_index, $any_array_ref)));
    }
    return -1;
}

sub array_find {
    # JavaScript-like Array.find() function array_find_v4
    my ($callback_function_ref, $any_array_ref) = @_;
    for (my $array_item_index = 0; ($array_item_index < scalar(@{$any_array_ref})); $array_item_index += 1) {
        my $array_item = $any_array_ref->[$array_item_index];
        return $array_item if (boolean($callback_function_ref->($array_item, $array_item_index, $any_array_ref)));
    }
    return undef;
}

sub array_includes {
    # JavaScript-like Array.includes() function array_includes_v4
    my ($search_element, $any_array_ref) = @_;
    for (my $array_item_index = 0; ($array_item_index < scalar(@{$any_array_ref})); $array_item_index += 1) {
        my $array_item = $any_array_ref->[$array_item_index];
        return (1 == 1) if (is_triple_equal($array_item, $search_element));
    }
    return (1 == 0);
}

sub array_map {
    # JavaScript-like Array.map() function array_map_v2
    my ($callback_function_ref, $any_array_ref) = @_;
    my $new_array_ref = [];
    for (my $array_item_index = 0; ($array_item_index < scalar(@{$any_array_ref})); $array_item_index += 1) {
        my $array_item = $any_array_ref->[$array_item_index];
        push(@{$new_array_ref}, ($callback_function_ref->($array_item, $array_item_index, $any_array_ref)));
    }
    return $new_array_ref;
}

sub array_some {
    # JavaScript-like Array.some() function array_some_v4
    my ($callback_function_ref, $any_array_ref) = @_;
    for (my $array_item_index = 0; ($array_item_index < scalar(@{$any_array_ref})); $array_item_index += 1) {
        my $array_item = $any_array_ref->[$array_item_index];
        return (1 == 1) if (boolean($callback_function_ref->($array_item, $array_item_index, $any_array_ref)));
    }
    return (1 == 0);
}

sub object_entries {
    # JavaScript-like Object.entries() function
    my ($any_object_ref) = @_;
    my $new_array_ref = [];
    my $object_entry_index = 0;
    while (my ($object_key, $object_value) = each(%{$any_object_ref})) {
        push(@{$new_array_ref}, [$object_key, $object_value]);
        $object_entry_index += 1;
    }
    return $new_array_ref;
}

sub pipe_v2 {
    my $pipe_last_result_ref = undef;
    my $pipe_result_ref = array_reduce((sub {
        my ($current_result_ref, $current_argument_ref) = @_;
        $pipe_last_result_ref = $current_result_ref;
        return $current_argument_ref if (is_triple_equal(get_type($current_result_ref), throw_error_if_null($js_like_type_ref->{"Null"})));
        return $current_argument_ref->($current_result_ref) if (is_triple_equal(get_type($current_argument_ref), throw_error_if_null($js_like_type_ref->{"Function"})));
        return undef;
    }), \@_, undef);
    return $pipe_result_ref->($pipe_last_result_ref) if (is_triple_equal(get_type($pipe_result_ref), throw_error_if_null($js_like_type_ref->{"Function"})));
    return $pipe_result_ref;
}

console_log("\n# JavaScript-like Spread Syntax (...) in Perl");

my $fruits_ref = ["Mango", "Melon", "Banana"];
console_log(string_interpolation("fruits: ", json_stringify($fruits_ref)));

my $vegetables_ref = ["Carrot", "Tomato"];
console_log(string_interpolation("vegetables: ", json_stringify($vegetables_ref)));

my $country_capitals_in_asia_ref = {
    "Thailand" => "Bangkok",
    "China" => "Beijing",
    "Japan" => "Tokyo"
};
console_log(string_interpolation("country_capitals_in_asia: ", json_stringify($country_capitals_in_asia_ref, "pretty" => (1 == 1))));

my $country_capitals_in_europe_ref = {
    "France" => "Paris",
    "England" => "London"
};
console_log(string_interpolation("country_capitals_in_europe: ", json_stringify($country_capitals_in_europe_ref, "pretty" => (1 == 1))));

console_log("\n# [...array1, ...array2]:\n");

my $combination1_ref = [(@{$fruits_ref}, @{$vegetables_ref})];
console_log(string_interpolation("combination1: ", json_stringify($combination1_ref, "pretty" => (1 == 1))));
# combination1: [
#     "Mango",
#     "Melon",
#     "Banana",
#     "Carrot",
#     "Tomato"
# ]

my $combination2_ref = [(@{$fruits_ref}, @{["Cucumber", "Cabbage"]})];
console_log(string_interpolation("combination2: ", json_stringify($combination2_ref, "pretty" => (1 == 1))));
# combination2: [
#     "Mango",
#     "Melon",
#     "Banana",
#     "Cucumber",
#     "Cabbage"
# ]

console_log("\n# { ...object1, ...object2 }:\n");

my $combination3_ref = {(%{$country_capitals_in_asia_ref}, %{$country_capitals_in_europe_ref})};
console_log(string_interpolation("combination3: ", json_stringify($combination3_ref, "pretty" => (1 == 1))));
# combination3: {
#     "Thailand": "Bangkok",
#     "China": "Beijing",
#     "Japan": "Tokyo",
#     "France": "Paris",
#     "England": "London"
# }

my $combination4_ref = {(%{$country_capitals_in_asia_ref}, %{{"Germany" => "Berlin", "Italy" => "Rome"}})};
console_log(string_interpolation("combination4: ", json_stringify($combination4_ref, "pretty" => (1 == 1))));
# combination4: {
#     "Thailand": "Bangkok",
#     "China": "Beijing",
#     "Japan": "Tokyo",
#     "Germany": "Berlin",
#     "Italy": "Rome"
# }

console_log("\n# [...array1, array2] || [...array1, newArrayItem1, newArrayItem2]:\n");

my $combination5_ref = [(@{$fruits_ref}, $vegetables_ref)];
console_log(string_interpolation("combination5: ", json_stringify($combination5_ref, "pretty" => (1 == 1))));
# combination5: [
#     "Mango",
#     "Melon",
#     "Banana",
#     [
#         "Carrot",
#         "Tomato"
#     ]
# ]

my $combination6_ref = [(@{$fruits_ref}, ["Cucumber", "Cabbage"])];
console_log(string_interpolation("combination6: ", json_stringify($combination6_ref, "pretty" => (1 == 1))));
# combination6: [
#     "Mango",
#     "Melon",
#     "Banana",
#     [
#         "Cucumber",
#         "Cabbage"
#     ]
# ]

console_log("\n# [...array1, object1] || [...array1, newArrayItem1, newArrayItem2]:\n");

my $combination7_ref = [(@{$fruits_ref}, $country_capitals_in_asia_ref)];
console_log(string_interpolation("combination7: ", json_stringify($combination7_ref, "pretty" => (1 == 1))));
# combination7: [
#     "Mango",
#     "Melon",
#     "Banana",
#     {
#         "Thailand": "Bangkok",
#         "China": "Beijing",
#         "Japan": "Tokyo"
#     }
# ]

my $combination8_ref = [(@{$fruits_ref}, {"Germany" => "Berlin", "Italy" => "Rome"} )];
console_log(string_interpolation("combination8: ", json_stringify($combination8_ref, "pretty" => (1 == 1))));
# combination8: [
#     "Mango",
#     "Melon",
#     "Banana",
#     {
#         "Germany": "Berlin",
#         "Italy": "Rome"
#     }
# ]

console_log("\n# { ...object1, object2 } || { ...object1, objectKey: objectValue }:\n");

my $combination9_ref = {(%{$country_capitals_in_asia_ref}, "country_capitals_in_europe" => $country_capitals_in_europe_ref)};
console_log(string_interpolation("combination9: ", json_stringify($combination9_ref, "pretty" => (1 == 1))));
# combination9: {
#    "Thailand" : "Bangkok",
#    "China" : "Beijing",
#    "Japan" : "Tokyo",
#    "country_capitals_in_europe" : {
#       "France" : "Paris",
#       "England" : "London"
#    }
# }

my $combination10_ref = {(%{$country_capitals_in_asia_ref}, "country_capitals_in_europe" => {"Germany" => "Berlin", "Italy" => "Rome"})};
console_log(string_interpolation("combination10: ", json_stringify($combination10_ref, "pretty" => (1 == 1))));
# combination10: {
#     "Thailand": "Bangkok",
#     "China": "Beijing",
#     "Japan": "Tokyo",
#     "country_capitals_in_europe": {
#         "Germany": "Berlin",
#         "Italy": "Rome"
#     }
# }

console_log("\n# { ...object1, array2 } || { ...object1, objectKey: objectValue }:\n");

my $combination11_ref = {(%{$country_capitals_in_asia_ref}, "vegetables" => $vegetables_ref)};
console_log(string_interpolation("combination11: ", json_stringify($combination11_ref, "pretty" => (1 == 1))));
# combination11: {
#     "Thailand": "Bangkok",
#     "China": "Beijing",
#     "Japan": "Tokyo",
#     "vegetables": [
#         "Carrot",
#         "Tomato"
#     ]
# }

my $combination12_ref = {(%{$country_capitals_in_asia_ref}, "vegetables" => ["Cucumber", "Cabbage"])};
console_log(string_interpolation("combination12: ", json_stringify($combination12_ref, "pretty" => (1 == 1))));
# combination12: {
#     "Thailand": "Bangkok",
#     "China": "Beijing",
#     "Japan": "Tokyo",
#     "vegetables": [
#         "Cucumber",
#         "Cabbage"
#     ]
# }

console_log("\n# { ...object1, ...array2 }:\n");

my $combination13_ref = {(%{$country_capitals_in_asia_ref}, %{object_from_entries(array_entries($vegetables_ref))})};
console_log(string_interpolation("combination13: ", json_stringify($combination13_ref, "pretty" => (1 == 1))));
# combination13: {
#    "Thailand" : "Bangkok",
#    "China" : "Beijing",
#    "Japan" : "Tokyo",
#    "0" : "Carrot",
#    "1" : "Tomato"
# }

my $combination14_ref = {(%{$country_capitals_in_asia_ref}, %{object_from_entries(array_entries(["Cucumber", "Cabbage"]))})};
console_log(string_interpolation("combination14: ", json_stringify($combination14_ref, "pretty" => (1 == 1))));
# combination14: {
#    "Thailand" : "Bangkok",
#    "China" : "Beijing",
#    "Japan" : "Tokyo",
#    "0" : "Cucumber",
#    "1" : "Cabbage"
# }

# console_log("\n# [...array1, ...object1]: # this combination throw an error in JavaScript\n");

# # this combination throw an error in JavaScript
# my $combination_error_in_javascript1_ref = [(@{$fruits_ref}, %{$country_capitals_in_asia_ref})];
# console_log(string_interpolation("combination_error_in_javascript1: ", json_stringify($combination_error_in_javascript1_ref, "pretty" => (1 == 1))));

# # this combination throw an error in JavaScript
# my $combination_error_in_javascript2_ref = [(@{$fruits_ref}, %{{"country_capitals_in_europe" => {"Germany" => "Berlin", "Italy" => "Rome"}}})];
# console_log(string_interpolation("combination_error_in_javascript2: ", json_stringify($combination_error_in_javascript2_ref, "pretty" => (1 == 1))));
